src/leader.c

```diff
@@ -450,3 +450,49 @@ int leader__barrier(struct leader *l, struct barrier *barrier, barrier_cb cb)
 	}
 	return 0;
 }
+
+int leader__reset(struct leader *l, struct exec *exec, exec_cb cb)
+{
+	tracef("leader reset");
+	sqlite3_vfs *vfs;
+	dqlite_vfs_frame *frames;
+	unsigned n, i;
+	int rv;
+
+	exec->leader = l;
+	memset(&exec->barrier, 0, sizeof(exec->barrier));
+	exec->stmt = NULL;
+	exec->cb = cb;
+	l->exec = exec;
+
+	rv = sqlite3_db_config(l->conn, SQLITE_DBCONFIG_RESET_DATABASE, 1, 0);
+	assert(rv == 0);
+	exec->status = sqlite3_exec(l->conn, "VACUUM", NULL, NULL, NULL);
+	rv = sqlite3_db_config(l->conn, SQLITE_DBCONFIG_RESET_DATABASE, 0, 0);
```

Comment:
Out of curiosity, does dqlite's memory footprint shrink after resetting a DB? If it's not shrinking, we should look into it.

Reply:
Does the `VACUUM` actually succeed?

Reply:
> Hmm, I don't see any change in number of DB pages, number of WAL frames, or number of shm regions after leaderApplyFrames returns vs. before `VACUUM`. @freeekanayaka, can you think of anything that would be preventing our VFS from dropping allocated memory when SQLite is done using it? (I could also just be looking in the wrong places -- I don't fully understand the VFS subsystem yet.)

Not on the top of my head, but I also don't have a clear picture of the changes that are being made. The SQLite memory allocator has some helper functions to track the amount of allocated memory, perhaps that could be useful to check.